r/src/arrow_rcpp.h

```diff
@@ -63,124 +68,222 @@ struct ns {
 };
 
-}  // namespace traits
+}  // namespace r
+}  // namespace arrow
 
-namespace internal {
+namespace cpp11 {
 
 template <typename T>
-inline SEXP wrap_dispatch(const T& x, Rcpp::traits::wrap_type_shared_ptr_tag);
+using enable_if_shared_ptr = typename std::enable_if<
+    std::is_same<std::shared_ptr<typename T::element_type>, T>::value, T>::type;
 
 template <typename T>
-inline SEXP wrap_dispatch(const T& x, Rcpp::traits::wrap_type_unique_ptr_tag);
+enable_if_shared_ptr<T> as_cpp(SEXP from) {
+  return arrow::r::ConstRefSmartPtrInput<typename T::element_type, std::shared_ptr>(from);
+}
 
-}  // namespace internal
-}  // namespace Rcpp
+template <typename T>
+SEXP as_sexp(const std::shared_ptr<T>& ptr) {
+  return Rcpp::XPtr<std::shared_ptr<T>>(new std::shared_ptr<T>(ptr));
+}
 
-#include <Rcpp.h>
+template <typename T>
+cpp11::list as_sexp(const std::vector<std::shared_ptr<T>>& vec) {
+  R_xlen_t n = vec.size();
+  SEXP res = PROTECT(Rf_allocVector(VECSXP, n));
+  for (R_xlen_t i = 0; i < n; i++) {
+    SET_VECTOR_ELT(res, i, as_sexp(vec[i]));
+  }
+  UNPROTECT(1);
+  return res;
+}
 
-namespace Rcpp {
-namespace internal {
+template <typename E, typename std::enable_if<std::is_enum<E>::value>::type* = nullptr>
+SEXP as_sexp(E e) {
+  return as_sexp(static_cast<int>(e));
+}
 
-template <typename T>
-inline SEXP wrap_dispatch(const T& x, Rcpp::traits::wrap_type_shared_ptr_tag) {
-  return Rcpp::XPtr<std::shared_ptr<typename T::element_type>>(
-      new std::shared_ptr<typename T::element_type>(x));
+}  // namespace cpp11
+
+namespace arrow {
+namespace r {
+
+template <typename T, typename r_vec, typename Lambda>
+r_vec to_r_vector(const std::vector<std::shared_ptr<T>>& x, Lambda lambda) {
+  auto n = x.size();
+  r_vec out(n);
+  for (int i = 0; i < n; i++) {
+    out[i] = lambda(x[i]);
+  }
+  return out;
 }
 
-template <typename T>
-inline SEXP wrap_dispatch(const T& x, Rcpp::traits::wrap_type_unique_ptr_tag) {
-  return Rcpp::XPtr<std::unique_ptr<typename T::element_type>>(
-      new std::unique_ptr<typename T::element_type>(const_cast<T&>(x).release()));
+template <typename T, typename Lambda>
+cpp11::writable::strings to_r_strings(const std::vector<std::shared_ptr<T>>& x,
+                                      Lambda lambda) {
+  return to_r_vector<T, cpp11::writable::strings, Lambda>(x, lambda);
 }
 
-}  // namespace internal
+template <typename T>
+cpp11::writable::list to_r_list(const std::vector<std::shared_ptr<T>>& x) {
+  auto lambda = [](const std::shared_ptr<T>& t) { return cpp11::as_sexp(t); };
+  return to_r_vector<T, cpp11::writable::list, decltype(lambda)>(x, lambda);
+}
 
-}  // namespace Rcpp
+template <typename T, typename Lambda>
+cpp11::writable::list to_r_list(const std::vector<std::shared_ptr<T>>& x, Lambda lambda) {
+  auto lambda1 = [lambda](const std::shared_ptr<T>& t) {
+    return cpp11::as_sexp(lambda(t));
+  };
+  return to_r_vector<T, cpp11::writable::list, decltype(lambda1)>(x, lambda1);
+}
```

Comment:
The declared lambda does not escape this scope, so capture by reference is acceptable
```suggestion
template <typename T, typename ToListElement>
cpp11::writable::list to_r_list(const std::vector<std::shared_ptr<T>>& x, ToListElement&& to_element) {
  auto as_sexp = [&](const std::shared_ptr<T>& t) {
    return cpp11::as_sexp(to_element(t));
  };
  return to_r_vector<cpp11::writable::list>(x, as_sexp);
}
```